scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/daemon/ReplaceAfterTerminateScheduler.java

```diff
@@ -15,18 +15,37 @@
 package com.amazonaws.blox.scheduling.scheduler.engine.daemon;
 
 import com.amazonaws.blox.scheduling.scheduler.engine.SchedulingAction;
+import com.amazonaws.blox.scheduling.state.ClusterSnapshot;
+import com.google.common.collect.ImmutableList;
 import java.util.List;
 import java.util.stream.Collectors;
 
 public class ReplaceAfterTerminateScheduler extends DaemonScheduler {
   public static final String ID = "ReplaceAfterTerminate";
 
   public List<SchedulingAction> schedule(DaemonEnvironment env, ClusterSummary summary) {
-    return summary
-        .getInstances()
-        .stream()
-        .filter(i -> env.hasMatchingTask(summary.tasksForInstance(i)))
-        .map(i -> env.startTaskFor(i))
-        .collect(Collectors.toList());
+    List<ClusterSnapshot.Task> tasksToStop =
+        summary
+            .getInstances()
+            .stream()
+            .flatMap(i -> summary.tasksForInstance(i).stream().filter(env::taskToStop))
+            .collect(Collectors.toList());
+
+    List<SchedulingAction> stopTaskActions =
+        tasksToStop.stream().map(env::stopTaskFor).collect(Collectors.toList());
+
+    List<SchedulingAction> startTaskActions =
+        summary
+            .getInstances()
+            .stream()
+            .filter(i -> env.needsToAssign(summary.tasksForInstance(i)))
+            .filter(i -> env.noRunningTasks(summary.tasksForInstance(i)))
```

Comment:
Any reason for splitting out `needsToAssign` and `noRunningTasks`? It doesn't look like they're ever used separately. Perhaps combining them into `isMissingHealthyTask(tasks)` would make this a bit clearer?

Reply:
Sure.

Reply:
After thinking through the logic, we actually only need to check noRunningTasks() for start. 
If the task is HEALTHY, even if it belongs to another environment/ another version, we still won't start a new task. We'll just wait for the scheduler to stop this current task.
Therefore, we only need to check the status of the task, regardless of the environment name and task version.

Reply:
> If the task is HEALTHY, even if it belongs to another environment/ another version, we still won't start a new task

That doesn't sound right; we *do* want to start a daemon task on an instance, even if there's already unrelated (i.e. from another env/no env) tasks running on it. Am I missing something here?

Reply:
I see! I was under the impression that the same instance won't be shared across different environments. If there can be multiple environments on an instance, then yes, we do need the check on matching environment name.

Reply:
The correct change should be, just check the matching environment name and RUNNING task status. We don't need to check task version, as we don't want to start a task when an older version is still running: we will stop it first.